frontends/p4/simplifyDefUse.cpp

```diff
@@ -726,10 +729,21 @@ class RemoveUnused : public Transform {
                 auto mce = se.nodeWithSideEffect->to<IR::MethodCallExpression>();
                 return new IR::MethodCallStatement(statement->srcInfo, mce);
             }
+            // removing
             return new IR::EmptyStatement();
         }
         return statement;
     }
+    const IR::Node* postorder(IR::MethodCallStatement* mcs) override {
+        if (!hasUses->hasUses(getOriginal())) {
+            if (SideEffects::hasSideEffect(mcs->methodCall, refMap, typeMap)) {
```

Comment:
Why can't you just reuse the existing check method?

Reply:
I have tried with check method, but `T f( in T a)` function didn't removed.

Reply:
`bit<32> f(out bit<32> a) {
    a = 32w0;
    return a;
}
bit<32> g(in bit<32> a) {
    return a;
}
control c(out bit<32> x, out bit<32> y) {
    action simple_action() {
        y = g(x);
        f(x);
        g(x);
    }
    apply {
        hdr h = { 0 };
        bool b = h.isValid();
        b = h.isValid();
        b = h.isValid();
        b = h.isValid();
        h.isValid();
        h.isValid();
        h.isValid();
        h.setValid();
        if (b) {
            x = h.a;
            x = h.a;
            f(h.a);
            f(h.a);
            g(h.a);
            g(h.a);
        } else {
            x = f(h.a);
            x = g(h.a);
            f(h.a);
        }
        simple_action();
    }
}` after SimplifyDefUse pass 
`bit<32> f(out bit<32> a) {
    a = 32w0;
    return a;
}
bit<32> g(in bit<32> a) {
    return a;
}
control c(out bit<32> x, out bit<32> y) {
    @name("h") hdr h_0;
    @name("b") bool b_0;
    @name("simple_action") action simple_action_0() {
        y = g(x);
        f(x);
        ;
    }
    apply {
        h_0 = (hdr){a = 32w0};
        ;
        ;
        ;
        b_0 = h_0.isValid();
        ;
        ;
        ;
        h_0.setValid();
        if (b_0) {
            ;
            x = h_0.a;
            f(h_0.a);
            f(h_0.a);
            ;
            ;
        } else {
            x = f(h_0.a);
            x = g(h_0.a);
            f(h_0.a);
        }
        simple_action_0();
    }
}
`

Reply:
Not all functions with only 'in' arguments will have no side-effects.

Reply:
Can you give me an example?

Reply:
log_msg from v1model

Reply:
You are right! 
Are we only sure for isValid() that it doesn't  produce side effects?

Reply:
There are "standard" methods, like isValid, for which the compiler can make assumptions.
Anything that is "extern" we should play it safe. 
Some annotations were added to the spec for this purpose; they are used in the compiler in some places, but not consistently everywhere (and certainly not here). See this section: https://p4.org/p4-spec/docs/P4-16-v1.2.1.html#sec-extern-annotations. So if you combine your test for 'in' arguments with this annotation the result should be safe.

Reply:
What do you suggest to change my code in that way, or use existing method check?
I will definitely take a look at that section.

Reply:
The SideEffects visitor does a deep traversal of the expression. That is necessary if you use it before the SideEffects pass. After SideEffects an expression has at most one operation with side-effects.
We should document when the visitor is safe to use. 
Ideally we should build some set of functions that can be reused for checking for side-effects, today many passes reimplement their own. 
Whatever you build, it must be "safe" for the intended use cases.

Reply:
Thank you for the explanation. I'm going to make some changes now.